### DISCUSSION TITLE
ActionScript worker in iOS adhoc not working

### DISCUSSION THREAD

**Discussion by myflashlab** [airsdk/Adobe-Runtime-Support#2175] [General]
I feel like I'm doing something wrong but I can't figure out what. I have setup the worker and it works perfectly when I am debugging on a real iOS device but as soon as I build the app in adhoc mode, the worker does nothing. the worker state shows `running` though but I get no reaction from the worker.

Since this is happening only in adhoc mode only, I cannot see any error logs whatsoever. Any ideas?

a little bit of more information:
I am compiling the working swf file like this:
```xml
    <target name="compile worker">
        <exec executable="${MXMLC}" failonerror="true">
            <arg line="
                        -debug=false
                        -source-path+=${worker.src}
                        -source-path+=${worker.src2}
                        -source-path+=${worker.src3}
                        -library-path+=${AS3_CORE_LIB_SWC}
                        -external-library-path+=${AIR_SWC}
                        -swf-version=${swfVersion}
                        -output ${worker.swf}
                        ${worker.documentClass}
            "/>
        </exec>
    </target>
```

And I am embedding the worker .swf file like this:
```
[Embed(source="../../workers/worker.swf", mimeType="application/octet-stream")]
private static var Worker_ByteClass:Class;
```

I also made sure that the main app .swf and worker .swf file are not built in debug mode `-debug=false` and they worked fine if compiling like `-package ipa-test-interpreter` but they don't work when using `-package ipa-ad-hoc` or `-package ipa-app-store`

**Comment by myflashlab** [comment]
And I have this problem with the latest SDK and older one. so, I don't think it's an SDK problem but maybe I have missed something 🤔

**Reply by ajwfrost** [reply]
I think the way you're embedding the SWF, it will be embedded as the full/raw SWF file containing all the ActionScript code .. but this can't then run on an iOS device (except in the debug-interpreter mode) because Apple don't allow scripts (other than their own).

You'd need to load in that worker SWF dynamically but first prepare it to strip out the AS3 code which then gets compiled into the main iOS application, via the `<iPhone><externalSwfs>` application descriptor entry..

**Reply by myflashlab** [reply]
Correct. that was exactly the problem. I had that piece of code used on desktop apps for a long time and had totally forgotten that iOS does not allow that. thanks for the reminder.

However, I didn't use `<externalSwfs>` thing because I didn't want to load the swf from a URL. instead, I packaged the worker.swf along with the main swf at the root of iOS and then loaded the worker swf like:

```
		var loader : URLLoader = new URLLoader();
		loader.dataFormat = URLLoaderDataFormat.BINARY;
		loader.addEventListener(Event.COMPLETE, onComplete);
		loader.addEventListener(IOErrorEvent.IO_ERROR, onError);
		loader.addEventListener(SecurityErrorEvent.SECURITY_ERROR, onError);
		loader.load(new URLRequest("worker.swf"));
```
And when loaded ar available, I feed them into the worker and all worked good :)

**Reply by ajwfrost** [reply]
Great, thanks -> it's a while since I actually did that workaround, I couldn't quite remember what the best way is for bringing in a separate worker swf. I tend to find the simplest thing is just have one SWF/class and get the behaviour to diverge in their initialisation code based on the `isPrimordial` flag...

**Reply by myflashlab** [reply]
> the simplest thing is just have one SWF/class and

For test reasons, this solution works and it's easy. However, if your main swf file is big, then it will take a long time for the worker to start which was not a good thing in our case. to make sure the worker is initiated as fast as possible, compiling it in a separated .swf file is best.

**Reply by kasparv** [reply]
> Correct. that was exactly the problem. I had that piece of code used on desktop apps for a long time and had totally forgotten that iOS does not allow that. thanks for the reminder.
> 
> However, I didn't use `<externalSwfs>` thing because I didn't want to load the swf from a URL. instead, I packaged the worker.swf along with the main swf at the root of iOS and then loaded the worker swf like:
> 
> ```
> 		var loader : URLLoader = new URLLoader();
> 		loader.dataFormat = URLLoaderDataFormat.BINARY;
> 		loader.addEventListener(Event.COMPLETE, onComplete);
> 		loader.addEventListener(IOErrorEvent.IO_ERROR, onError);
> 		loader.addEventListener(SecurityErrorEvent.SECURITY_ERROR, onError);
> 		loader.load(new URLRequest("worker.swf"));
> ```
> 
> And when loaded ar available, I feed them into the worker and all worked good :)

@myflashlab could you also show how you're initializing the worker?

I'm trying to initialize it like this
```
private function onLoadComplete(event:Event):void
		{
			worker = WorkerDomain.current.createWorker(event.target.data as ByteArray, true);

			resultChannel = worker.createMessageChannel(Worker.current);
			resultChannel.addEventListener(Event.CHANNEL_MESSAGE, onIncomingMessage);
			worker.setSharedProperty("resultChannel", resultChannel);

			commandChannel = Worker.current.createMessageChannel(worker);
			worker.setSharedProperty("commandChannel", commandChannel);
			worker.addEventListener(Event.WORKER_STATE, onWorkerStateChange);
			worker.start();
		}
```
		
_but I get two `onWorkerStateChange` events with `worker.state = "terminated"` on iOS, on other platforms everything works fine..._

Edit:
I did find a solution to my issue - had to omit the `.swf` from the included swf file as the `adt` package process seems to strip it into much smaller file. Removing file type did the trick and the included worker is functioning correctly!

Is there a way to add swf files to the .ipa package root without adt scrambling them?

**Reply by myflashlab** [reply]
@kasparv yeah, I remember that double `terminated` problem and I hate how easy it was to be fixed :) I have written the solution in a comment in my app so I'd never forget that again. here's my comment which I wrote after many many hours of pulling my hair

```
		// IMPORTANT: on iOS adhoc, the above simulator MUST be commented out because even if it is compiled, in adhoc 
		// mode duplicated classes cannot merge and will cause in double worker terminations
```

So, in my code, to be able to easily debug the operations in workers, I had a helper function which was initializing the classes which are normally only available in the worker. When building for release, that helper function was never called and the worker worked instead. This was find in all platforms except for adhoc. I understood that eventhough the helper func is not called, the classes are imported inside the main .swf file anyway and in adhoc builds, for some reason, duplicated classes cannot merge and the worker failes. That's why :) I hope this saves time from others seeing that unexpected worker termination.

@ajwfrost You might find the above explanation helpful if it would make any difference in how the classes are merged in adhoc?

**Reply by kasparv** [reply]
@myflashlab thanks for the reply! Do you use single swf and initialize worker only with code? 
I compile the worker swf separately before I compile the .ipa but after I found the workaround to remove file type `.swf` from the file name, I get one `onWorkerStateChange` with state `running` but I don't seem to get any other communication working between main swf and the worker...
@myflashlab does your Easy AS Workers project work with iOS?
